app/components/form/fields/TlsCertsField.tsx

```diff
@@ -0,0 +1,137 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Copyright Oxide Computer Company
+ */
+import { useState } from 'react'
+import type { Control } from 'react-hook-form'
+import { useController } from 'react-hook-form'
+
+import { Button, Error16Icon, FieldLabel, MiniTable, Modal } from '@oxide/ui'
+
+import { DescriptionField, FileField, NameField } from 'app/components/form'
+import type { SiloCreateInput } from 'app/forms/silo-create'
+import { useForm } from 'app/hooks'
+import { readBlobAsText } from 'app/util/file'
+
+export function TlsCertsField({ control }: { control: Control<SiloCreateInput> }) {
+  const [showAddCert, setShowAddCert] = useState(false)
+
+  const {
+    field: { value: items, onChange },
+  } = useController({ control, name: 'tlsCertificates' })
+
+  return (
+    <>
+      <div className="max-w-lg">
+        <FieldLabel id="tls-certificates-label" className="mb-3">
+          TLS Certificates
+        </FieldLabel>
+        {!!items.length && (
+          <MiniTable.Table className="mb-4">
+            <MiniTable.Header>
+              <MiniTable.HeadCell>Name</MiniTable.HeadCell>
+              {/* For remove button */}
+              <MiniTable.HeadCell className="w-12" />
+            </MiniTable.Header>
+            <MiniTable.Body>
+              {items.map((item, index) => (
+                <MiniTable.Row
+                  tabIndex={0}
+                  aria-rowindex={index + 1}
+                  aria-label={`Name: ${item.name}, Description: ${item.description}`}
+                  key={item.name}
+                >
+                  <MiniTable.Cell>{item.name}</MiniTable.Cell>
+                  <MiniTable.Cell>
+                    <button
+                      onClick={() => onChange(items.filter((i) => i.name !== item.name))}
+                    >
+                      <Error16Icon title={`remove ${item.name}`} />
+                    </button>
+                  </MiniTable.Cell>
+                </MiniTable.Row>
+              ))}
+            </MiniTable.Body>
+          </MiniTable.Table>
+        )}
+
+        <Button size="sm" onClick={() => setShowAddCert(true)}>
+          Add TLS certificate
+        </Button>
+      </div>
+
+      {showAddCert && (
+        <AddCertModal
+          onDismiss={() => setShowAddCert(false)}
+          onSubmit={async (values) => {
+            const keypair = {
+              cert: await readBlobAsText(values.cert),
+              key: await readBlobAsText(values.key),
+            }
+            onChange([...items, { ...values, ...keypair }])
```

Comment:
`onChange` accepts any here, do we want to do `[...items. {...values, ...keypair}] as TlsCertificate[]` to get a typecheck on this.

Reply:
I bet there’s a way to get a better type on `onChange`, will try it later.

Reply:
Or not. I tried.

```ts
const certCreate: (typeof items)[number] = {
  ...values,
  cert: await values.cert.text(),
  key: await values.key.text(),
}
onChange([...items, certCreate])
```

Nothing I tried felt perfect, but this feels pretty good — only assert that the value matches the items in `items`. This is very close to your idea, except 

a) `as` is a cast, so it doesn't really check. `satisfies` would work better in that spot, and
b) because we're inferring from the type of `items`, the type of `items` can change and we will only see an error if `certCreate` doesn't match it